lib/upload-manager.js

```diff
@@ -0,0 +1,287 @@
+		this.canceled = false;
+	}
+
+	/**
+	 * Get the final object representation of this chunk for the API
+	 * @returns {UploadPart} The chunk object
+	 */
+	getData() {
+
+		return this.data;
+	}
+
+	/**
+	 * Upload a chunk to the API
+	 * @returns {void}
+	 */
+	upload() {
+
+		this.client.files.uploadPart(this.sessionID, this.chunk, this.offset, this.totalSize, null, (err, data) => {
+
+			if (this.canceled) {
+				this.chunk = null;
+				return;
+			}
+
+			if (err) {
+				// handle the error or retry
+				if (err.statusCode) {
+					// an API error, probably not retryable!
+					this.emit('error', err);
+				} else {
+					// maybe a network error, retry
+					this.retry = setTimeout(() => this.upload(), RETRY_INTERVAL_MS);
+				}
+				return;
+			}
+
+			// Record the chunk data for commit, and try to free up the chunk buffer
+			this.data = data;
+			this.chunk = null;
+			this.emit('uploaded', data);
+		});
+	}
+
+	/**
+	 * Cancel trying to upload a chunk, preventing it from retrying and clearing
+	 * the associated buffer
+	 * @returns {void}
+	 */
+	cancel() {
+
+		clearTimeout(this.retry);
+		this.chunk = null;
+		this.canceled = true;
+	}
+}
+
+// -----------------------------------------------------------------------------
+// Public
+// -----------------------------------------------------------------------------
+
+//@TODO: rename to something better
+/** Manager for uploading a file in chunks */
+class UploadManager extends EventEmitter {
+
+	/**
+	 * Create an upload manager
+	 * @param {BoxClient} client The client to use to upload the file
+	 * @param {Object} uploadSessionInfo The upload session info to use for chunked upload
+	 * @param {ReadableStream|Buffer|string} file The file to upload
+	 * @param {int} size The size of the file to be uploaded
+	 * @param {Object} [options] Optional parameters
+	 */
+	constructor(client, uploadSessionInfo, file, size, options) {
+
+		super();
+
+		this.client = client;
+		this.sessionID = uploadSessionInfo.upload_session_id;
+		this.partSize = uploadSessionInfo.part_size;
+		this.uploadSessionInfo = uploadSessionInfo;
+
+		if (file instanceof ReadableStream) {
+			// Pause the stream so we can read specific chunks from it
+			this.stream = file.pause();
+			this.streamBuffer = [];
+		} else if (file instanceof Buffer || typeof file === 'string') {
+			this.file = file;
+		} else {
+			throw new TypeError('file must be a Stream, Buffer, or string!');
+		}
+
+		this.size = size;
+		this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+
+		this.isStarted = false;
+		this.numChunksInFlight = 0;
+		this.chunks = [];
+		this.position = 0;
+		this.fileHash = crypto.createHash('sha1');
+	}
+
+	/**
+	 * Start an upload
+	 * @returns {void}
+	 */
+	start() {
+
+		if (this.isStarted) {
+			return;
+		}
+
+		// Create the initial chunks
+		for (let i = 0; i < this.options.parallelism; i++) {
+			this._getNextChunk(chunk => (chunk ? this._uploadChunk(chunk) : this._commit()));
```

Comment:
I think it should be fine, since the very last chunk to be uploaded must emit the `'uploaded'` event, which first decrements the number of chunks in flight to zero and then triggers the `_commit()` call.